tools/terraform/fleetdm_client/fleetdm_client.go

```diff
@@ -0,0 +1,276 @@
+package fleetdm_client
+
+// This file gives us a nice API to use to call FleetDM's API. It's focused
+// only on teams.
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+)
+
+var prefix = "/api/v1/fleet"
+var teamPrefix = prefix + "/teams"
+
+type Team struct {
+	Name         string      `json:"name"`
+	ID           int64       `json:"id"`
+	Description  string      `json:"description"`
+	AgentOptions interface{} `json:"agent_options"`
+	Scripts      interface{} `json:"scripts"`
+	Secrets      []struct {
+		Secret  string    `json:"secret"`
+		Created time.Time `json:"created_at"`
+		TeamID  int       `json:"team_id"`
+	} `json:"secrets"`
+}
+
+type TeamCreate struct {
+	Name        string `json:"name,omitempty"`
+	Description string `json:"description,omitempty"`
+}
+
+type TeamGetResponse struct {
+	Team struct {
+		Team
+	} `json:"team"`
+}
+
+type TeamQueryResponse struct {
+	Teams []Team `json:"teams"`
+}
+
+// FleetDMClient is a FleetDM HTTP client that overrides http.DefaultClient.
+type FleetDMClient struct {
+	*http.Client
+	URL    string
+	APIKey string
+}
+
+// NewFleetDMClient creates a new instance of FleetDMClient with the provided
+// URL and API key.
+func NewFleetDMClient(url, apiKey string) *FleetDMClient {
+	return &FleetDMClient{
+		Client: http.DefaultClient,
+		URL:    url,
+		APIKey: apiKey,
+	}
+}
+
+// Do will add necessary headers and call the http.Client.Do method.
+func (c *FleetDMClient) do(req *http.Request, query string) (*http.Response, error) {
+	// Add the API key to the request header
+	req.Header.Add("Authorization", fmt.Sprintf("Bearer %s", c.APIKey))
+	req.Header.Add("Accept", `application/json`)
+	// Set the request URL based on the client URL
+	req.URL, _ = url.Parse(c.URL + req.URL.Path)
+	if query != "" {
+		req.URL.RawQuery = query
+	}
+	// Send the request using the embedded http.Client
+	return c.Client.Do(req)
+}
+
+// TeamNameToId will return the ID of a team given the name.
+func (c *FleetDMClient) TeamNameToId(name string) (int64, error) {
+	req, err := http.NewRequest(http.MethodGet, teamPrefix, nil)
+	if err != nil {
+		return 0, fmt.Errorf("failed to create GET request for %s: %w", teamPrefix, err)
+	}
+	query := fmt.Sprintf("query=%s", name)
```

Comment:
This will not encode the query properly for names with unicode or special chars in it (the query end up in the `RawQuery` field of the `URL` value, which expects properly query-encoded data, e.g. https://go.dev/play/p/HMnJspG5GXz).

You need to create an `url.Values` and add values to it, then get its encoded value, something like:
```
vals := make(url.Values)
vals.Add("query", name)
query := vals.Encode()
```

Reply:
(or of course change the signature of `do` to receive `url.Values` instead, and encode it when setting the `RawQuery` field)

Reply:
Good catch; I think that taking the query in do() as a string and then encoding it into req.URL works well.

Reply:
I put the Encode() inside of do()